### ISSUE TITLE
Support long file/path names for TarWriter/TarReader

### ISSUE THREAD

**Issue by ghost** [rubygems/rubygems#1376]
Currently [`TarWriter`](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/package/tar_writer.rb#L311) limits file/path names to 256 (or so) characters.

However, for example this [test file](https://www.dropbox.com/s/vfodcqa26neqz9m/test.tar?dl=0) can contain files with path names exceeding 256 characters. That file was made by `GNU tar 1.28`.

Would you please support long file name (path name) for `TarWriter` (and possibly `TarReader`)?

Thanks,

**Comment by djberg96** [comment]
I'm not actually sure why that check is in there unless it's some attempt to comply with an old POSIX spec.

**Comment by drbrain** [comment]
Is there a specification for modern tar files?

Also, old rubygems won't be able to install tar files if we make this upgrade.

**Comment by drbrain** [comment]
I suppose we could add support and a warning upon packaging when using long file names, but we need a link to the spec to implement support.

**Comment by djberg96** [comment]
@drbrain This is the best I could find:

https://www.gnu.org/software/tar/manual/html_section/tar_68.html
https://en.wikipedia.org/wiki/Tar_(computing)

I think the 256 path limit is POSIX-1988 or earlier that hung around in some implementations. I vaguely recall Solaris tar suffering from this issue and having to replace it with GNU tar. From what I can gather, most Unixy implementations set path max at 4k now, or have no limit. Now if we're talking about a *filename* limit, that I think it still generally set to 256 on Unixy platforms, but it depends on the filesystem type.

Anyway, what happens if we just remove that check?

**Comment by drbrain** [comment]
New tar stores the filename differently, the 256 bytes comes from the sizes of the `name` and `prefix` entries in the [tar header](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/package/tar_header.rb#L9-L26)

In order to support longer filenames we need to implement this `‼` thing, it seems:

```
$ hexdump -C test.tar | more
00000000  21 21 30 2f 00 00 00 00  00 00 00 00 00 00 00 00  |!!0/............|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000060  00 00 00 00 30 30 30 30  37 30 30 00 30 30 30 31  |....0000700.0001|
00000070  37 35 30 00 30 30 30 31  37 35 30 00 30 30 30 30  |750.0001750.0000|
00000080  30 30 30 30 30 30 30 00  31 32 36 31 37 33 32 33  |0000000.12617323|
00000090  33 36 30 00 30 30 37 30  35 34 00 20 35 00 00 00  |360.007054. 5...|
000000a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000100  00 75 73 74 61 72 20 20  00 68 62 00 00 00 00 00  |.ustar  .hb.....|
00000110  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000120  00 00 00 00 00 00 00 00  00 68 62 00 00 00 00 00  |.........hb.....|
00000130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000200  21 21 30 2f 31 2f 00 00  00 00 00 00 00 00 00 00  |!!0/1/..........|
00000210  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000260  00 00 00 00 30 30 30 30  37 30 30 00 30 30 30 31  |....0000700.0001|
00000270  37 35 30 00 30 30 30 31  37 35 30 00 30 30 30 30  |750.0001750.0000|
00000280  30 30 30 30 30 30 30 00  31 32 36 31 37 33 32 33  |0000000.12617323|
00000290  33 36 30 00 30 30 37 32  31 34 00 20 35 00 00 00  |360.007214. 5...|
000002a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000300  00 75 73 74 61 72 20 20  00 68 62 00 00 00 00 00  |.ustar  .hb.....|
00000310  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000320  00 00 00 00 00 00 00 00  00 68 62 00 00 00 00 00  |.........hb.....|
00000330  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000400  21 21 30 2f 31 2f 32 32  2f 00 00 00 00 00 00 00  |!!0/1/22/.......|
00000410  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000460  00 00 00 00 30 30 30 30  37 30 30 00 30 30 30 31  |....0000700.0001|
00000470  37 35 30 00 30 30 30 31  37 35 30 00 30 30 30 30  |750.0001750.0000|
00000480  30 30 30 30 30 30 30 00  31 32 36 31 37 33 32 33  |0000000.12617323|
00000490  33 36 30 00 30 30 37 34  33 37 00 20 35 00 00 00  |360.007437. 5...|
000004a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000500  00 75 73 74 61 72 20 20  00 68 62 00 00 00 00 00  |.ustar  .hb.....|
00000510  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000520  00 00 00 00 00 00 00 00  00 68 62 00 00 00 00 00  |.........hb.....|
00000530  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000600  21 21 30 2f 31 2f 32 32  2f 33 33 33 2f 00 00 00  |!!0/1/22/333/...|
00000610  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
```

**Comment by haibison** [comment]
Thank you for accepting this request, that's great  :-)

I was going to find tar format from Wikipedia and GNU tar project, like Daniel did.

That folder name `!!0` was my mistake, it has no special meaning. I often prefix folder/file names with `!!` and tell Git to ignore those names.

For backward compatibility, I think you can add a flag like `use_long_names`, default to `false`. That way people won't need to update code...

**Comment by haibison** [comment]
I used monkey patching to remove the limit check, it produced a [broken file](https://www.dropbox.com/s/9cbnskj47q7km6c/test2.tar?dl=0).

This is code:

```ruby
#!/usr/bin/env ruby

require 'rubygems/package'

IO_BUFFER = 1024 * 64

class Gem::Package::TarWriter
    def split_name(name) # :nodoc:
        if name.bytesize <= 100 then
            prefix = ""
        else
            parts = name.split(/\//)
            newname = parts.pop
            nxt = ""

            loop do
                nxt = parts.pop
                break if newname.bytesize + 1 + nxt.bytesize > 100
                newname = nxt + "/" + newname
            end

            prefix = (parts + [nxt]).join "/"
            name = newname
        end

        return name, prefix
    end
end

if __FILE__ == $0
    tar_reader, tar_writer = IO.pipe

    # tar handler
    tar_handler_thread = Thread.new do
        File.open('test2.tar', 'wb') do |io|
            while data = tar_reader.read(IO_BUFFER) do
                io.write data
            end
        end
    end

    Gem::Package::TarWriter.new(tar_writer) do |writer|
        original_size = 0

        0.upto(50) do |index|
            path = '!!0'
            1.upto(index) { |i| path = File.join(path, i.to_s * i) }

            test = "Hellboy III, please!\n"
            writer.add_file_simple(File.join(path, 'test.txt'), 0644, test.bytesize) do |io|
                io.write test
                original_size += test.bytesize
            end
        end

        puts " > Total size = #{original_size} bytes"
    end

    tar_writer.close
    tar_handler_thread.join
    tar_reader.close
end
```

**Comment by drbrain** [comment]
Yes, it would, because you're overflowing the 512 byte tar header. Per the hexdump, modern tar stores the filename in a `!!` section, but without a specification its too hard to know if it is implemented correctly.

**Comment by haibison** [comment]
I'm sorry I accidentally deleted the sample file. But from the source, it seems that the limit is `256`...

Thanks,

**Comment by haines** [comment]
We just got bitten by this while using the docker-api gem which relies on the TarWriter. Any chance of revisiting this feature request?

The limit of 256-character paths and 100-character filenames is imposed by the ustar (POSIX.1-1988) format. By default, GNU tar uses its own format which supports unlimited filename lengths (see [here](https://www.gnu.org/software/tar/manual/tar.html#SEC134)); I suspect that was what the previous sample file was using. However, I think the appropriate format to use is pax (POSIX.1-2001) format, which is the default in BSD tar and will become the default in a future version of GNU tar.

The spec for the pax format is [here](http://pubs.opengroup.org/onlinepubs/009695399/utilities/pax.html#tag_04_100_13_01).

**Comment by grosser** [comment]
@bronzdoc can you reopen or is this solved ?

**Comment by chuckixia** [comment]
here's a quick way to repro:
`StringIO.open('') do |write_io|
  write_io.set_encoding('binary')
  Gem::Package::TarWriter.new(write_io) do |tar|
    contents = 'hello i am a small file'
    filename = 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
    tar.add_file_simple(filename, 0x444, contents.length) do |io|
      io.write(contents)
    end
  end
end`

**Comment by alexshpilkin** [comment]
The modern formats (POSIX-1988 ustar and its successor POSIX-2001 pax) are in POSIX as part of the [specification of the unfortunately stillborn pax utility][1], but the best writeup is probably the [tar(5)][2] man page from libarchive. The default is ustar with GNU extensions for GNU tar and pax for BSD tar.

I actually can’t get GNU tar (1.32) to produce the `!!` things from @drbrain’s comment, only GNU-style L entries or POSIX-style `path=` keys in x entries. Same for BSD tar (3.3.3) under Linux.

[1]: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html
[2]: https://github.com/libarchive/libarchive/wiki/ManPageTar5
[3]: http://schilytools.sourceforge.net/man/man4/star.4.html

**Comment by zenspider** [comment]
@alexshpilkin you might want to update your comment. The `tar(5)` wiki entry is gone. It seems to have been replaced with https://github.com/libarchive/libarchive/wiki/ManualPages and that points to https://www.freebsd.org/cgi/man.cgi?query=tar&sektion=5&format=html

**Comment by alexshpilkin** [comment]
@zenspider Done, thanks!

**Comment by rubyFeedback** [comment]
I've just hit the issue with a second project.

The annoying part is that the file name was a .yml file so not even a .rb file:

ERROR:  While executing gem ... (Gem::Package::TooLongFileName)
    File "lib/studium/yaml/curricula/awesome_components_that_have_to_be_intelligently_considered_when_doing_similar_upgrades_in_the_future.yml" has a too long name (should be 100 or less)

That name is indeed long but it works fine on my local linux machine.

From reading the thread, it seems to be related to tar/zlib? My biggest
problem is that I can not work around the issue. I can shorten the name
but I'd run into more and more similar problems the longer my local files
names are. I think it would be easier for me to just package this into a
.zip or .tar.xz and distribute this rather than to rely on gem+rubygems,
but ideally I'd like to continue using gem and rubygems. (I tend to also
distribute the yaml files that are part of a package).

For the time being, could the limit be increased by a bit? Ideally an
API to manually determine this would be nice too. Or some way to
specify to gem to ignore certain files. Right now this is causing 
me issues; I have to manually remove or rename the file at hand
before publishing, but this is a workaround, since the only reason
for me to rename it would be due to gem refusing to upload it..

**Comment by simi** [comment]
@rubyFeedback I think it is clear what's the problem and how it can affect common workflows.

:information_source: If I understand it well, the problem is how to tackle this in rubygems. If we update to another tar specification (for example the `pax` one), older rubygems will not be able to read (unpack) gems created by newer version and that's really huge breaking change in whole ecosystem.

:thinking: Potentially we can introduce new gem format (not having this path length limitation - for example by using `pax` tar format for data), keep both formats supported and add option to use newer format by new gemspec attribute (which will check also proper `required_rubygems_verions` is set to only rubygems supporting this new format). We will keep old format the default one for backward compatibility and once we reach time when all supported rubygems versions can handle the new format, we should be able to switch to new format by default.

:face_with_head_bandage:  If you would like to still distribute your long filepaths along your gem, I can imagine you can create ruby build extension (triggered on gem install) to unpack some archive during install time (for example using minitar, rubyzip or so specified as dependency). Anyway that's still a bandaid.

**Comment by zenspider** [comment]
I do not see why this requires a new gem format or gemspec attributes at all.

**Comment by haines** [comment]
pax is largely backwards-compatible with ustar. 

For the majority of gems (including all existing gems), there would be no impact of extracting a version packaged with pax using a ustar-based rubygems, because the archives produced would be identical if the gem contains no long filenames (or non-ASCII filenames).

For new gems that do contain such filenames, those files would be extracted to the wrong location, because the ustar-based rubygems would ignore the extended header entries and just take the filename as written on the file's standard header.

But for those gems, couldn't we just use the `required_rubygems_version` in the gemspec to avoid that problem?

**Comment by simi** [comment]
@zenspider by new format I meant new tar format of `data.tar.gz` in gem file. Whole idea about the new attribute was based on check of `required_rubygems_verions` to be inline with new format and fail install on older rubygems.

@haines Does that mean you would like to check for long filepaths during packaging and enforce (or just warn about) `required_rubygems_version` should be set to some version, otherwise gem will not be installable properly (in case it is not set already)? That actually sounds quite simpler than my idea. :+1:

**Comment by haines** [comment]
@simi yep, I reckon that would work. It should probably be an error rather than a warning.

**Comment by simi** [comment]
:+1: @haines I think error will be the best as well (to actually break gem push build on CI for example on problems).
